core-processor/src/executor.rs

```diff
@@ -287,90 +287,102 @@ pub fn execute_wasm<A: ProcessorExt + EnvExt + IntoExtInfo + 'static, E: Environ
+                program_id,
+                &mut pages_initial_data,
+                static_pages,
+                memory,
+            )
+        },
     ) {
-        return Err(ExecutionError {
-            program_id,
-            gas_amount: env.into_gas_amount(),
-            reason,
-        });
-    }
-
-    // Execute program in backend env.
-    let BackendReport { termination, info } = match env.execute(&kind, |mem| {
-        // released pages initial data will be added to `pages_initial_data` after execution.
-        if A::is_lazy_pages_enabled() {
-            A::lazy_pages_post_execution_actions(mem, &mut pages_initial_data)
-        } else {
-            Ok(())
+        Ok(BackendReport {
+            termination_reason: termination,
+            memory_wrap: memory,
+            stack_end_page,
+        }) => {
+            // released pages initial data will be added to `pages_initial_data` after execution.
+            if A::is_lazy_pages_enabled() {
+                if let Err(e) =
+                    A::lazy_pages_post_execution_actions(&memory, &mut pages_initial_data)
+                {
+                    return Err(ExecutionError {
+                        program_id,
+                        gas_amount: ext.into_gas_amount(),
+                        reason: ExecutionErrorReason::Backend(e.to_string()),
+                    });
+                }
+            }
+            (termination, memory, stack_end_page)
         }
-    }) {
-        Ok(report) => report,
+
         Err(e) => {
             return Err(ExecutionError {
                 program_id,
-                gas_amount: e.gas_amount.clone(),
+                gas_amount: ext.into_gas_amount(),
                 reason: ExecutionErrorReason::Backend(e.to_string()),
             })
         }
     };
 
+    // Page which is right after stack last page
+    log::trace!("Stack end page = {:?}", stack_end_page);
+
     log::debug!("Termination reason: {:?}", termination);
 
-    // Parsing outcome.
-    let kind = match termination {
-        TerminationReason::Exit(value_dest) => DispatchResultKind::Exit(value_dest),
-        TerminationReason::Leave | TerminationReason::Success => DispatchResultKind::Success,
-        TerminationReason::Trap(explanation) => {
-            log::debug!(
-                "💥 Trap during execution of {}\n📔 Explanation: {}",
-                program_id,
-                explanation,
-            );
+    match ext.into_ext_info(&memory, stack_end_page.unwrap_or_default()) {
+        Ok(info) => {
+            // Parsing outcome.
+            let kind = match termination {
+                TerminationReason::Exit(value_dest) => DispatchResultKind::Exit(value_dest),
+                TerminationReason::Leave | TerminationReason::Success => {
+                    DispatchResultKind::Success
+                }
+                TerminationReason::Trap(explanation) => {
+                    log::debug!(
```

Comment:
What the reason to increase code nesting so much here?
It will be much more readable like this:
```rust
let info = match ext.into_ext_info(...).ok_or_else(|| ... )?;
// here logic with `info`
```